### PR TITLE
feat: SlippageProtection Use Fill Adjustment in Phase2 of routing.

### PR DESCRIPTION
**Pull request by dekz**
# Description

https://github.com/0xProject/protocol/pull/480

* Re-exported `contract-addresses` from `asset-swapper` to avoid a footgun of different versions
* Added a `FillAdjustor` using the `SlippageModelManager`
* Added `enableSlippageProtection` for integrators to signal whether they want slippage adjusted routing based on the slippage model
### Simbots
* [enableSlippageProtection=false](https://metabase.spaceship.0x.org/dashboard/186?run_id=disabled-slippage-protect&adjusted_amount_bought_win_tolerance__bps_=5) should observe very little difference as the changes are disabled.
* [enableSlippageProtection=true]()

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Test changes on the staging environment with [Matcha API Staging](https://api-staging.matcha.xyz).

    -   [ ] SRA/Limit orders
    -   [ ] Swap endpoints
    -   [ ] Meta transaction endpoints
    -   [ ] Depth charts

    For more information see `0x API Matcha smoke test runbook` in Quip.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -138,7 +138,6 @@
         "@0x/api-utils": "^0.0.2",
         "@0x/assert": "^3.0.34",
         "@0x/asset-swapper": "^16.60.1",
-        "@0x/contract-addresses": "^6.14.0",
```
Comment (dekz) [inline package.json:141]: Exported through asset-swapper to avoid loading different versions.

File: src/services/swap_service.ts [hunk 2]
```diff
@@ -294,7 +294,7 @@ export class SwapService {
                 ...rfqt,
                 intentOnFilling: rfqt && rfqt.intentOnFilling ? true : false,
                 integrator: integrator!,
-                makerEndpointMaxResponseTimeMs: RFQT_REQUEST_MAX_RESPONSE_MS,
+                makerEndpointMaxResponseTimeMs: 15000 || RFQT_REQUEST_MAX_RESPONSE_MS,
```
Comment (dekz) [inline src/services/swap_service.ts:297]:
```suggestion
                makerEndpointMaxResponseTimeMs: RFQT_REQUEST_MAX_RESPONSE_MS,
```

File: src/services/swap_service.ts [hunk 3]
```diff
@@ -329,6 +329,14 @@ export class SwapService {
             shouldGenerateQuoteReport,
             shouldIncludePriceComparisonsReport: !!includePriceComparisons,
             samplerMetrics: SAMPLER_METRICS,
+            fillAdjustor: this._slippageModelManager
```
Comment (dekz) [inline src/services/swap_service.ts:332]: We will also flip here on an API param

File: src/utils/price_comparison_utils.ts [hunk 4]
```diff
@@ -20,7 +20,7 @@ import { SourceComparison } from '../types';
 
 // NOTE: Our internal Uniswap gas usage may be lower than the Uniswap UI usage
 // Therefore we need to adjust the gas estimates to be representative of using the Uniswap UI.
-const gasScheduleWithOverrides: FeeSchedule = {
+const gasScheduleWithOverrides: GasSchedule = {
```
Comment (dekz) [inline src/utils/price_comparison_utils.ts:23]: Renamed the type in Asset-swapper to be clearer

File: src/config.ts [hunk 5]
```diff
@@ -222,7 +222,7 @@ export const DB_ORDERS_UPDATE_CHUNK_SIZE = 300;
 export const ETHEREUM_RPC_URL = assertEnvVarType('ETHEREUM_RPC_URL', process.env.ETHEREUM_RPC_URL, EnvVarType.UrlList);
 // Timeout in seconds to wait for an RPC request (default 5000)
 export const RPC_REQUEST_TIMEOUT = _.isEmpty(process.env.RPC_REQUEST_TIMEOUT)
-    ? 5000
+    ? 10000
```
Comment (dekz) [inline src/config.ts:225]:
```suggestion
    ? 5000
```

File: src/handlers/swap_handlers.ts [hunk 6]
```diff
@@ -224,7 +225,13 @@ export class SwapHandlers {
         );
         if (params.includePriceComparisons && quote.priceComparisonsReport) {
             const side = params.sellAmount ? MarketOperation.Sell : MarketOperation.Buy;
-            const priceComparisons = priceComparisonUtils.getPriceComparisonFromQuote(CHAIN_ID, side, quote);
+            const priceComparisons = priceComparisonUtils.getPriceComparisonFromQuote(
+                CHAIN_ID,
+                side,
+                quote,
+                this._swapService.slippageModelManager,
```
Comment (dekz) [inline src/handlers/swap_handlers.ts:232]: Adding the `expectedSlippage` in the `priceComparisons`, mostly for quick insight/debugging.

File: src/services/swap_service.ts [hunk 7]
```diff
@@ -333,6 +334,15 @@ export class SwapService {
             shouldGenerateQuoteReport,
             shouldIncludePriceComparisonsReport: !!includePriceComparisons,
             samplerMetrics: SAMPLER_METRICS,
+            fillAdjustor:
+                enableSlippageProtect && this.slippageModelManager
```
Comment (dekz) [inline src/services/swap_service.ts:338]: if the API param is passed and enabled, or the default is to enable it globally, we use a `SlippageModelFillAdjustor`. 

If not we just default to the `IdentityFillAdjustor` which makes no modifications.

File: src/services/swap_service.ts [hunk 8]
```diff
@@ -444,10 +454,10 @@ export class SwapService {
             }
         }
         // If any sources can be undeterministic in gas costs, we add a buffer
-        const hasUndeterministicFills = _.flatten(swapQuote.orders.map((order) => order.fills)).some((fill) =>
-            [ERC20BridgeSource.Native, ERC20BridgeSource.MultiBridge].includes(fill.source),
+        const hasUndeterministicFills = swapQuote.orders.some((order) =>
+            [ERC20BridgeSource.Native, ERC20BridgeSource.MultiBridge].includes(order.source),
         );
-        const undeterministicMultiplier = hasUndeterministicFills ? GAS_LIMIT_BUFFER_MULTIPLIER : 1;
+        const undeterministicMultiplier = isQuote && hasUndeterministicFills ? GAS_LIMIT_BUFFER_MULTIPLIER : 1;
```
Comment (dekz) [inline src/services/swap_service.ts:460]: Only add this particular buffer on real live quotes, not the price endpoint.

File: src/utils/sllippage_model_fill_adjustor.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                this.maxSlippageRate,
+            );
+
+            if (expectedSlippage === null) {
+                return f;
+            }
+
+            const fill = { ...f };
+            // 1000 * (1 + -0.02938) = 970
+            const slippageAdjustedOutput = f.output
```
Comment (novaliu86) [inline src/utils/sllippage_model_fill_adjustor.ts:32]: Shall the `slippageAdjustedOutput` depend on `side`? Seems the adjust will always decrease `output` amount as `expectedSlippage` is mostly negative. But for sell the output is the higher the better, while for buy the output is the lower the better for takers.

**Review by novaliu86 — APPROVED** [review]
LGTM

**Comment by dekz** [comment]
deploy production

**Comment by dekz** [comment]
deploy production

**Comment by dekz** [comment]
deploy production

**Pull request merged**